v8/usr/man/man1/ps.1
.TH PS 1
.SH NAME
ps \- process status
.SH SYNOPSIS
.B ps
[
.BR aFfhlnrtux number
] ...
.SH DESCRIPTION
.I Ps
prints information about processes.
Normally, only the user's
non-shell processes are eligible for printing; specifying
.B a
includes other user's processes; specifying
.B x
includes shell processes as well.  (A stopped shell is printed even if the
.B x
flag is missing.)
.PP
All output formats include, for each process, the process id, its
control terminal, status, cpu time, and command.
The status is given
by a sequence of letters:
.ta \w'W'u+1m +\w'Idle (> 20 seconds).'u+3m +\w'W'u+1m
.PP
R	Running.	T	Stopped.
.br
P	Waiting for a page.	S	Sleeping (< 20 seconds).
.br
I	Idle (> 20 seconds).
.br
W	Swapped out.	N	Process is `nice'; see \f2nice\f1(1).
.PP
R, T, P, S, and I are mutually exclusive.
.PP
Options are given in blank-separated clusters:
.nr xx \w'\f2number'u+1m/1n
.TP \n(xx
.B a
give information about all users (ordinarily
only one's own processes are displayed).
.PD 0
.TP
.BI F file
display all processes having an open
.IR file ;
this option must be the last in a cluster.
.TP
.B f
display file descriptors (with prepended process id).
.TP
.B ff
display file descriptors (omitting process id).
.TP
.B h
print column headers.
.TP
.B l
(long) print the process size,
resident set size, wait channel, and parent process id.
.TP
.B n
print unsorted
(output is normally sorted by process id).
.TP
.B r
print processes which match the current real or effective user id
(normally only processes matching the current effective user id
are printed).
.TP
.BI t x
print processes whose controlling tty is
.I x,
where
.I x
is either ``.'' (for user's tty), or one of the forms printed by
.I ps,
e.g.
.I t3
for /dev/tty3,
.I tconso
for console,
.I tpt12
for /dev/pt/pt12,
.I t?
for processes with no tty, etc. This option must be the last in a cluster.
.TP
.B u
include a name of the
effective userid and cpu share in percent; sort by usage.
.TP
.B x
include shell and
.I init
processes.
.TP
.I number
print the designated process;
must be the last option in a cluster.
.PD
.PP
Multiple
.BR F ", " t ", and"
.I number
options are allowed; the union of all selections is printed.
.SH FILES
.ta \w'/usr/net/friends'u+1m
/proc	process images
.br
/dev/drum	swap device
.br
/dev/kmemr	kernel memory
.br
/dev	searched to find tty names
.br
/dev/pt	searched to find pt names
.br
/dev/dk	searched to find dk names
.br
/etc/fstab	searched to find local file system names
.br
/usr/net/friends	searched to find network file system names
.SH "SEE ALSO"
kill(1), nice(2), proc(4), oops(8)
.SH BUGS
Things can change while
.I ps
is running.
.br
A process may alter its apparent command name and arguments.


v8/usr/src/cmd/ps/ps.c
#include <stdio.h>
#include "ps.h"

struct proc proc;

union { struct user user; char userchars[UPAGES*NBPG]; } usp;
#define u usp.user
#define upages usp.userchars

static char tmpstr[4096]; static Xflag;

main(argc, argv)
int argc; char **argv;
{
	register struct direct *dp;
	register Psline *ps; Psline *psbegin, *psend;
	register char *cp; register ncp, fd;

	progname=argv[0];

	if ((devlist = getdir("/dev", devlist)) == 0)
		error("can't getdir /dev");
	if ((devlist = getdir("/dev/dk", devlist)) == 0)
		error("can't getdir /dev/dk");
	if ((devlist = getdir("/dev/pt", devlist)) == 0)
		error("can't getdir /dev/pt");
	if ((drum = open("/dev/drum", 0)) < 0)
		error("can't open /dev/drum");

	myuid = getuid(); mypid = getpid();

	if (argc > 1)
		selbeg = selend = Malloc(Select, argc-1);
	doarg(*++argv), Xflag = xflag && !tflag;
	while (--argc > 1)
		doarg(*++argv);

	if (hflag) {
		if (uflag) printf("User    %%cpu ");
		if (lflag) printf("size  rss  ppid wchan ");
		printf("  pid tty   stat  time command\n");
	}

	if (Nflag && !tflag && !Fflag) {
		register Select *sp;
		Nflag = tflag = 0;
		++aflag; ++xflag;
		for (sp=selbeg; sp<selend; sp++) {
			if ((fd = doselect(sp->id, &proc, &u)) < 0)
				continue;
			if (printp(tmpstr, fd, &proc, &u) > tmpstr)
				printf("%s", tmpstr);
		}
		return 0;
	}

	if ((prlist = getdir("/proc", 0)) == 0)
		error("can't getdir /proc");
	if ((psbegin = Malloc(Psline, prlist->end - prlist->begin)) == 0)
		error("can't malloc ps line table");

	for (ps = psbegin, dp = prlist->begin; dp < prlist->end; dp++) {
		if (dp->d_ino == 0 || dp->d_name[0] == '.')
			continue;
		if ((fd = doselect(dp->d_name, &proc, &u)) < 0)
			continue;
		if ((cp = printp(tmpstr, fd, &proc, &u)) <= tmpstr)
			continue;
		if (nflag) {
			printf("%s", tmpstr);
			continue;
		}
		ncp = cp - tmpstr + 1;
		ps->string = memcpy(malloc(ncp), tmpstr, ncp);
		ps->weight = proc.p_pid;
		if (uflag)
			ps->weight += (1001 - (int)(1000.0*proc.p_pctcpu+0.5))<<16;
		ps++;
	}
	if (!nflag) {
		ptsort(psbegin, psend = --ps);
		for (ps=psbegin; ps<=psend; ps++)
			printf("%s", ps->string);
	}
	exit(0);
}

doarg(cp)
register char *cp;
{
	struct stat fds;
	register Select *sp;
	xflag = 0;
	if (cp) do switch (*cp) {
		case 0:
			cp = 0; break;
		case '-':
			break;
		case 't':
			if (*++cp == '.') {
				if (fstat(0, &fds) == 0) {
					cp = gettty(fds.st_ino);
					if (*cp == '?')
						cp = 0;
				} else {
					fprintf(stderr, "0: cannot fstat\n");
					cp = 0;
				}
			}
			if (cp) {
				++tflag;
				sp = selend++;
				sp->flag = SELTTY;
				if (Xflag || xflag)
					sp->flag |= SELXFL;
				sp->id = cp;
				cp = 0;
			}
			break;
		case 'a':
			aflag++; break;
		case 'F':
			if (stat(++cp, &fds) == 0) {
				Fflag++;
				sp = selend++;
				sp->flag = SELFIL;
				sp->id = cp;
				sp->dev = fds.st_dev;
				sp->ino = fds.st_ino;
				if (kernel <= 0)
					kernel = open("/dev/kmemr", 0);
			} else
				fprintf(stderr, "%s: cannot stat\n", cp);
			cp = 0;
			break;
		case 'f':
			if (kernel <= 0)
				kernel = open("/dev/kmemr", 0);
			if (kernel)
				fflag++, getfstab();
			break;
		case 'h':
			hflag++; break;
		case 'l':
			lflag++; break;
		case 'n':
			nflag++; break;
		case 'r':
			rflag++; break;
		case 'u':
			uflag++; getlogins(); break;
		case 'x':
			xflag++; break;
		case 'T':
			Tflag++; break;
		default:
			if (*cp < '0' || *cp > '9')
				usage();
			++Nflag;
			sp = selend++;
			sp->flag = SELXFL;
			sp->id = cp;
			cp = 0;
	} while (cp++);
	return 0;
}

error(s)
char *s;
{
	fprintf(stderr, "%s: %s\n", progname, s);
	exit(1);
}

usage()
{
	fprintf(stderr, "Usage: %s [afFhlnrtux#] ...\n", progname);
	exit(1);
}
